File: petclinic-backend/src/main/java/victor/training/petclinic/mapper/PetTypeMapper.java
package victor.training.petclinic.mapper;

import org.springframework.stereotype.Component;
import victor.training.petclinic.domain.PetType;
import victor.training.petclinic.rest.dto.PetTypeDto;
import victor.training.petclinic.rest.dto.PetTypeFieldsDto;

import java.util.ArrayList;
import java.util.List;

@Component
public class PetTypeMapper {

    public PetType toPetType(PetTypeFieldsDto petTypeFieldsDto) {
        PetType petType = new PetType();
        petType.setName(petTypeFieldsDto.getName());
        return petType;
    }

    public PetTypeDto toPetTypeDto(PetType petType) {
        PetTypeDto petTypeDto = new PetTypeDto();
        petTypeDto.setName(petType.getName());
        petTypeDto.setId(petType.getId());
        return petTypeDto;
    }

    public PetTypeFieldsDto toPetTypeFieldsDto(PetType petType) {
        PetTypeFieldsDto petTypeFieldsDto = new PetTypeFieldsDto();
        petTypeFieldsDto.setName(petType.getName());
        return petTypeFieldsDto;
    }

    public List<PetTypeDto> toPetTypeDtos(List<PetType> petTypes) {
        if (petTypes == null) {
            return List.of();
        }
        List<PetTypeDto> dtos = new ArrayList<>(petTypes.size());
        for (PetType petType : petTypes) {
            dtos.add(toPetTypeDto(petType));
        }
        return dtos;
    }
}

File: petclinic-backend-ts/src/pet-types/pet-type.mapper.ts
import { PetType } from './pet-type.entity';
import { PetTypeDto } from './dto/pet-type.dto';
import { PetTypeFieldsDto } from './dto/pet-type-fields.dto';

/**
 * Ported from victor.training.petclinic.mapper.PetTypeMapper (MapStruct).
 *
 * STATELESS plain functions — no Nest DI, no @Injectable — mirroring
 * MapStruct's stateless mappers. Controllers import these directly.
 */

/** MapStruct: toPetType — `id` is ignored (left undefined for INSERT). */
export function toPetType(fieldsDto: PetTypeFieldsDto): PetType {
  const petType = new PetType();
  petType.name = fieldsDto.name;
  return petType;
}

/** MapStruct: toPetTypeDto. */
export function toPetTypeDto(petType: PetType): PetTypeDto {
  const dto = new PetTypeDto();
  dto.id = petType.id;
  dto.name = petType.name ?? '';
  return dto;
}

/** MapStruct: toPetTypeFieldsDto. */
export function toPetTypeFieldsDto(petType: PetType): PetTypeFieldsDto {
  const dto = new PetTypeFieldsDto();
  dto.name = petType.name ?? '';
  return dto;
}

/** MapStruct: toPetTypeDtos. */
export function toPetTypeDtos(petTypes: PetType[]): PetTypeDto[] {
  return petTypes.map(toPetTypeDto);
}
